crates/esthri/src/blocking.rs

```diff
@@ -36,52 +32,12 @@ pub async fn upload<T>(
     bucket: impl AsRef<str>,
     key: impl AsRef<str>,
     file: impl AsRef<Path>,
+    opts: &GenericOptParams,
```

Comment:
I think I would prefer to have a different OptParams type for each verb.  Right now you can do stuff like

```rust
let opts = GenericOptParamsBuilder::default().transparent_compression(true).build().unwrap();
let res = head_object(..., &opts);
```

This firstly doesn't make any sense but maybe more importantly doesn't help the user of the API understand which options are available for this verb.  As a user of the API, I would want to look at the definition of `GenericOptParams` here to understand which options are available for `head_object`.

Reply:
Initially I have gone down the path of different struct for different verb using respective param holders, however problems quickly arose, seeing that the lib crates has lots of interwoven verbs calling each other, e.g. sync function internally calling upload and download, or copy calling both upload and download etc, while these verbs are mostly consolidated into a single function, they end up with lots of shared parameters if they were to work with a single call

I think it's definitely a tradeoff is to have a shared struct where the members are well-known s3 related options like compression and delete. The other option is to have another go at diff struct for diff verb but need another round of refactoring to decouple these interwoven function. What do you think?

Reply:
If you implement `From` for easily transforming structs for some verbs into others does that help make it less convoluted?

If we had flattened or nested builders it might help out too but looks like this isn't implemented yet. 
 https://github.com/colin-kiegel/rust-derive-builder/issues/254

Reply:
I was thinking this same thing when I first looked over these changes: the need for different builder types will help with the overall clarity of the API. I agree that `From` is probably the best way to solve the interdependence issue, it'll require some boilerplate, but the gain in API clarity is worth it IMO.